Implemented frontend docs

simveit opened the pull request:
## Motivation

Rewrote frontend docs as jupyter notebook.

simveit committed:
Implemented frontend docs

simveit committed:
Adjust structure and removed markdown

simveit committed:
Duplicate patch to frontend folder to make CI run. Fix that later.

shuaills reviewed (changes requested):
Great job

simveit commented on docs/frontend/frontend.ipynb:
```diff
@@ -0,0 +1,491 @@
+    "from sglang.test.test_utils import is_in_ci\n",
+    "\n",
+    "if is_in_ci():\n",
+    "    from patch import launch_server_cmd\n",
```
Actually I am not sure if there is an easy way without using python project or `sys` to make `patch.py` accessible in all the subfolders (both of which IMO would be overly complex and make the notebooks less readable...)
Maybe @zhaochenyang20 has a better idea.
Also the CI still failed, it seemed the `!wget https://github.com/sgl-project/sglang/blob/main/test/lang/example_image.png?raw=true -O example_image.png` i use didn't work because after the image got not recognized. Any ideas about why that is?

zhaochenyang20 replied:
I agree with the current design of the patch. I mean including `patch.py` in sub-directories. I think for the image, you can refer to this https://docs.sglang.ai/backend/openai_api_vision.html

simveit replied:
ok i fixed the issue.
```python
import requests


def download_image(url, filename):
    response = requests.get(url)
    if response.status_code == 200:
        with open(filename, "wb") as f:
            f.write(response.content)
        print(f"Successfully downloaded {filename}")
    else:
        print(f"Failed to download image: Status code {response.status_code}")


image_url = "https://github.com/sgl-project/sglang/blob/main/test/lang/example_image.png?raw=true"
download_image(image_url, "example_image.png")
```
this code solved the problem

zhaochenyang20 committed:
Merge branch 'main' into feature/frontend-docs

simveit committed:
Try to fix problem with wget

simveit committed:
Merge branch 'main' into feature/frontend-docs

simveit committed:
Merge branch 'feature/frontend-docs' of github.com:simveit/sglang into feature/frontend-docs

simveit commented:
Before we merge this I will update the engine API now. I will put this also into this PR.

simveit committed:
Moved hidden states to examples, fixed vlm engine example, implemented vlm in engine doc, refactored frontend.

simveit commented:
@shuaills @zhaochenyang20 
I implemented vlm for engine doc, fixed mistake in example for engine with vlm (the script wants to access some tokenizer_manager attribute of engine which doesnt exist) and moved hidden_states extraction to examples.
For some reason the CI failed for the openai completions notebook which i didnt modify.
Shuai could you take a look at this please, i am not on computer now.
<img width="814" alt="Screenshot 2025-02-23 at 17 12 02" src="https://github.com/user-attachments/assets/ed985293-bfea-4e4d-bf26-f041f9fe9789" />

zhaochenyang20 commented:
@simveit I rerun it and let's see what will happen. Are you agree with the docs? @shuaills I can merge it

zhaochenyang20 committed:
Merge branch 'main' into feature/frontend-docs

zhaochenyang20 commented:
@simveit @shuaills I will wait approval from Shuai. But LGTM

shuaills commented on docs/backend/offline_engine_api.ipynb:
```diff
@@ -1,5 +1,16 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "\n",
+    "os.environ[\"CUDA_VISIBLE_DEVICES\"] = \"1\""
```
Remove this

simveit replied:
sure, sorry this was for my local dev setup.

shuaills commented on docs/backend/offline_engine_api.ipynb:
```diff
@@ -242,7 +257,42 @@
+    "image_token = conv.image_token\n",
+    "\n",
+    "# Convert image to bytes\n",
+    "image = Image.open(\"example_image.png\")\n",
```
Can we use url here? So we don't need to manage file system.

simveit replied:
i will adjust that to use the `load_image` function from `utils.py` (see below)

shuaills commented on examples/runtime/engine/offline_batch_inference_vlm.py:
```diff
@@ -5,56 +5,62 @@
+            f.write(response.content)
+        print(f"Successfully downloaded {filename}")
+    else:
+        print(f"Failed to download image: Status code {response.status_code}")
```
Download to file system may introduce some problems, can we use [url](https://docs.sglang.ai/backend/openai_api_vision.html#:~:text=image_url%22%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22image_url%22%3A%20%7B-,%22url%22%3A%20%22https%3A//github.com/sgl%2Dproject/sglang/blob/main/test/lang/example_image.png%3Fraw%3Dtrue%22,-%7D%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%5D%2C%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%5D%2C%0A%20%20%20%20max_tokens) instead?

simveit replied:
I don't believe that will make make a large difference. Under the hood something like 

```python
def load_image(image_file: Union[str, bytes]):
    from PIL import Image

    image = image_size = None

    if isinstance(image_file, bytes):
        image = Image.open(BytesIO(image_file))
    elif image_file.startswith("http://") or image_file.startswith("https://"):
        timeout = int(os.getenv("REQUEST_TIMEOUT", "3"))
        response = requests.get(image_file, timeout=timeout)
        image = Image.open(BytesIO(response.content))
    elif image_file.lower().endswith(("png", "jpg", "jpeg", "webp", "gif")):
        image = Image.open(image_file)
    elif image_file.startswith("data:"):
        image_file = image_file.split(",")[1]
        image = Image.open(BytesIO(base64.b64decode(image_file)))
    elif image_file.startswith("video:"):
        image_file = image_file.replace("video:", "")
        image, image_size = decode_video_base64(image_file)
    elif isinstance(image_file, str):
        image = Image.open(BytesIO(base64.b64decode(image_file)))
    else:
        raise ValueError(f"Invalid image: {image}")

    return image, image_size
```
will be called under the hood if we provide an url.
A syntax you use in this notebook was tried to use before for the example and it didn't run. (threw error).

simveit replied:
this will have especially benefit we don't save anything as a file and I agree that this is little bit undesirable.

shuaills replied:
The code mostly operates in memory, except when handling local file paths. But if we do `open(filename, "wb")` , it will accesse the file system.

simveit replied:
yes, i will adjust to use to utils function from sglang code base from above

simveit replied:
I fixed this @shuaills @zhaochenyang20 . The current implemetation is much easier and cleaner.

simveit committed:
Improved image handling

simveit committed:
Merge branch 'feature/frontend-docs' of github.com:simveit/sglang into feature/frontend-docs

simveit committed:
Moved sending request to getting started

simveit commented:
@zhaochenyang20 @shuaills i moved also the send request to the start and renamed section to "Getting started". Let me know if that is not what you intendend

zhaochenyang20 committed:
Merge branch 'main' into feature/frontend-docs

zhaochenyang20 commented:
@shuaills Thanks!

zhaochenyang20 commented:
<img width="966" alt="image" src="https://github.com/user-attachments/assets/72edb5f1-f12e-4079-af7b-4353738e4921" />

To me, i will remove `VLM Inference` and the get hidden state in these docs. But I will tell the users at the place I point to, "You can refer to these examples for [VLM offline inference](https://github.com/sgl-project/sglang/blob/main/examples/runtime/engine/offline_batch_inference_vlm.py) and [getting hidden states](xxx)."

For the getting hidden state examples, please make one, just following https://docs.sglang.ai/backend/offline_engine_api.html#Return-Hidden-States

@simveit

zhaochenyang20 commented on docs/index.rst:
```diff
@@ -12,15 +12,15 @@ The core features include:
+   :caption: Getting Started
 
    start/install.md
+   start/send_request.ipynb
```
if you move send_reqeusts here, the patch would fail and `send_request.ipynb` will take large VRAM. But I think you can add one `patch.py` here.

simveit replied:
i think we already have `patch.py` in start dir for this PR (see changed files)

simveit commented:
> <img alt="image" width="966" src="https://private-user-images.githubusercontent.com/74843776/417277633-72edb5f1-f12e-4079-af7b-4353738e4921.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA1OTUzNTksIm5iZiI6MTc0MDU5NTA1OSwicGF0aCI6Ii83NDg0Mzc3Ni80MTcyNzc2MzMtNzJlZGI1ZjEtZjEyZS00MDc5LWFmN2ItNDM1MzczOGU0OTIxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI2VDE4MzczOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFhMjlmMjc0OTBhMDFlZTkyNzBkOTUxNTM1NTBmZjllZTRiMDYwYTc5NGM5Y2UyYTBiZWY4ZDgzMjUwNjU5OTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yqPyA-NLZ90y4CKSltq8KOmI08XCWbaAMcD1krj1LaY">
> To me, i will remove `VLM Inference` and the get hidden state in these docs. But I will tell the users at the place I point to, "You can refer to these examples for [VLM offline inference](https://github.com/sgl-project/sglang/blob/main/examples/runtime/engine/offline_batch_inference_vlm.py) and [getting hidden states](xxx)."
> 
> For the getting hidden state examples, please make one, just following https://docs.sglang.ai/backend/offline_engine_api.html#Return-Hidden-States
> 
> @simveit

I both included hidden states example and also fixed the vlm example for engine in this PR. let me remove the example from notebook later and refer to the links

simveit committed:
Doc for hidden states, pointing to examples instead of including them

simveit committed:
fixup

simveit committed:
small fix

zhaochenyang20 committed:
Merge branch 'main' into feature/frontend-docs

simveit committed:
move back to backend

simveit committed:
Merge branch 'feature/frontend-docs' of github.com:simveit/sglang into feature/frontend-docs

simveit committed:
fix doctree

simveit committed:
fix doctree

zhaochenyang20 committed:
Merge branch 'main' into feature/frontend-docs
